encoding/ccf/decode.go

```diff
@@ -0,0 +1,2039 @@
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime/common"
+	cadenceErrors "github.com/onflow/cadence/runtime/errors"
+)
+
+// CBORDecMode
+//
+// See https://github.com/fxamacker/cbor:
+// "For best performance, reuse EncMode and DecMode after creating them."
+var CBORDecMode = func() cbor.DecMode {
+	decMode, err := cbor.DecOptions{
+		IntDec:           cbor.IntDecConvertNone,
+		MaxArrayElements: math.MaxInt64,
+		MaxMapPairs:      math.MaxInt64,
+		MaxNestedLevels:  math.MaxInt16,
+	}.DecMode()
+	if err != nil {
+		panic(err)
+	}
+	return decMode
+}()
+
+// A Decoder decodes CCF-encoded representations of Cadence values.
+type Decoder struct {
+	// CCF codec uses CBOR codec under the hood.
+	dec   *cbor.StreamDecoder
+	gauge common.MemoryGauge
+}
+
+// Decode returns a Cadence value decoded from its CCF-encoded representation.
+//
+// This function returns an error if the bytes represent CCF that is malformed,
+// invalid, or does not comply with requirements in the CCF specification.
+func Decode(gauge common.MemoryGauge, b []byte) (cadence.Value, error) {
+	dec := NewDecoder(gauge, b)
+
+	v, err := dec.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	if dec.dec.NumBytesDecoded() != len(b) {
+		return nil, cadenceErrors.NewDefaultUserError("ccf: failed to decode: decoded %d bytes, received %d bytes", dec.dec.NumBytesDecoded(), len(b))
+	}
+
+	return v, nil
+}
+
+// NewDecoder initializes a Decoder that will decode CCF-encoded bytes from the
+// given bytes.
+func NewDecoder(gauge common.MemoryGauge, b []byte) *Decoder {
+	// NOTE: encoded data is not copied by decoder.
+	// CCF codec uses CBOR codec under the hood.
+	return &Decoder{
+		dec:   CBORDecMode.NewByteStreamDecoder(b),
+		gauge: gauge,
+	}
+}
+
+// Decode reads CCF-encoded bytes and decodes them to a Cadence value.
+//
+// This function returns an error if the bytes represent CCF that is malformed,
+// invalid, or does not comply with requirements in the CCF specification.
+func (d *Decoder) Decode() (value cadence.Value, err error) {
+	// Capture panics that occur during decoding.
+	defer func() {
+		// Recover panic error if there is any.
+		if r := recover(); r != nil {
+			// Don't recover Go errors, internal errors, or non-errors.
+			switch r := r.(type) {
+			case goRuntime.Error, cadenceErrors.InternalError:
+				panic(r)
+			case error:
+				err = r
+			default:
+				panic(r)
+			}
+		}
+
+		// Add context to error if there is any.
+		if err != nil {
+			err = cadenceErrors.NewDefaultUserError("ccf: failed to decode: %s", err)
+		}
+	}()
+
+	// Decode top level message.
+	tagNum, err := d.dec.DecodeTagNumber()
+	if err != nil {
+		return nil, err
+	}
+
+	switch tagNum {
+	case CBORTagTypeDefAndValue:
+		// Decode ccf-typedef-and-value-message.
+		return d.decodeTypeDefAndValue()
+
+	case CBORTagTypeAndValue:
+		// Decode ccf-type-and-value-message.
+		return d.decodeTypeAndValue(newCadenceTypeByCCFTypeID())
+
+	default:
+		return nil, fmt.Errorf(
```

Comment:
It would be good to add tess for the uncovered error cases here and below (could be in a follow-up PR)

Reply:
Yes, more tests can be added in subsequent PRs for unhappy paths.